### PR TITLE
prune operation if it isn't needed

### PR DESCRIPTION
**Pull request by dollschasingmen**

### CODE REVIEW

**Comment by sritchie** [comment]
@dollschasingmen, sorry I never really looked at this. I know it's been a long, long time.

One thing we need to guarantee is that we include filtering variables in the set of variables that we protect, not just the final projection. a query like

```clojure
(<- [?x]
  (src ?x)
  (* ?x ?x :> ?square)
  (even? ?square))
```

Needs to return only even squares even though the filter variable's removed. Does that make sense? Let me know if you're still interested here and I can give you a super-fast review turnaround time. Thanks again!

**Comment by dollschasingmen** [comment]
@sritchie, np, thanks for not forgetting this PR :) 

This function doesn't prune filter predicates under any circumstances. The pruning action is limited to only non-filter and non-generator predicates. I added a few more tests to assert that (see latest commits).

Sorry, the function is a bit convoluted. It has been a long time and reading through it was not super simple even for me, ha. Not sure how to make it the logic more obvious/readable (suggestions very much appreciated here), but I  re-wrote the function and test descriptions to make it all a bit more clear (hopefully?)

File: cascalog-core/src/clj/cascalog/logic/parse.clj [hunk 1]
```diff
@@ -642,6 +642,32 @@ This won't work in distributed mode because of the ->Record functions."
...
 
+(defn prune-operations
+  "Remove non-generator & non-filter operations whose outvar(s) is not used, i.e., when the outvar(s)
+  do *not* intersect with the query's out-fields, generators outvars (for joins), invars of other operations,
+  and :sort option. Additionally, do not prune operations if a no-input operator exists"
+  [fields grouped options]
+  (let [check-input-operations (concat (grouped Operation)
+                                       (grouped FilterOperation)
+                                       (grouped Aggregator))]
+  (if (some #(empty? (:input %)) check-input-operations)
```
Comment (dollschasingmen) [inline cascalog-core/src/clj/cascalog/logic/parse.clj]: check for existance of no-input operators

File: cascalog-core/src/clj/cascalog/logic/parse.clj [hunk 2]
```diff
@@ -642,6 +642,32 @@ This won't work in distributed mode because of the ->Record functions."
...
+                                (concat fields)
+                                (concat (mapcat :fields generators))
+                                (concat (:sort options)))
+          pruned-operations (filter #(seq
+                                       (clojure.set/intersection
+                                            (set (:output %))
+                                            (set necessary-fields)))
+                                    (grouped Operation))]
+      (concat pruned-operations
+              (grouped FilterOperation))))))
```
Comment (dollschasingmen) [inline cascalog-core/src/clj/cascalog/logic/parse.clj]: return pruned-operations and filter operations. so basically, filter operations don't even get touched.

**Comment by sritchie** [comment]
This looks great. Currently compilation's forced at the subquery boundary, but I can see this advancing down the road to actually filter down across subqueries when the user does some projection like `(src ?x _)` when using a query.

I'm happy to merge this now, but I do have one more Q / suggestion; I might be wrong here, but reading the code it looks like this optimization might not be able to handle chains of operations that end in some unused variable, like

```clojure
(<- [?x]
  (src ?x)
  (* ?x ?x :> ?square)
  (dec ?square :> ?square-minus-one))
```

I bet the easiest way to deal with this would be to run the same optimization you're currently running, but to pass it through a `fixed-point` function like

```clojure
(defn fixed-point [f guess]
  (let [next (f guess)]
    (if (= guess next)
      next
      (fixed-point f next))))
```

and cull all operations backwards. Up to you if you think this is a good idea or worth doing in this pull request. I always feel lame asking for MORE work in a pull request, especially given how badly I dropped the ball on reviewing this one.

Thanks again!

**Comment by sritchie** [comment]
This is pretty closely related to 

https://github.com/nathanmarz/cascalog/issues/24

Both the chained-and-ignored predicate thing and duplicate operations happen a lot as a result of predicate macros.

**Comment by dollschasingmen** [comment]
yea - good point re: handling chains of operations. i think i got it work using basically that fixed-point idea, see latest commit. 

also agree about the subquery boundary and pred-macros -- sure would be nice to be able to prune into that as well :) but that, I think, I might prefer to leave to another PR.

**Comment by dollschasingmen** [comment]
hm not sure why CI failed, test passed locally

**Comment by sritchie** [comment]
Looks like a dependency download failure with cascalog-math. Thanks again!

**Comment by slpsys** [comment]
Sweet!

**Comment by sritchie** [comment]
Love that this works now:

```clojure
cascalog.api> (defn throw! [& xs] (throw (RuntimeException. "gotcha!")))
#'cascalog.api/throw!
cascalog.api> (??<- [?x] ([[1]] ?x) (throw! ?x :> ?ignored))
([1])
cascalog.api> (??<- [?x ?ignored] ([[1]] ?x) (throw! ?x :> ?ignored))
FlowException local step failed  cascading.flow.planner.FlowStepJob.blockOnJob (FlowStepJob.java:219)
```

**Pull request merged**